dbms/src/Interpreters/SystemLog.h

```diff
@@ -183,51 +176,86 @@ SystemLog<LogElement>::SystemLog(Context & context_,
 {
     log = &Logger::get("SystemLog (" + database_name + "." + table_name + ")");
 
-    data.reserve(DBMS_SYSTEM_LOG_QUEUE_SIZE);
-    saving_thread = ThreadFromGlobalPool([this] { threadFunction(); });
+    saving_thread = ThreadFromGlobalPool([this] { savingThreadFunction(); });
 }
 
 
 template <typename LogElement>
 void SystemLog<LogElement>::add(const LogElement & element)
 {
+    std::unique_lock lock(mutex);
+
     if (is_shutdown)
         return;
 
-    /// Without try we could block here in case of queue overflow.
-    if (!queue.tryPush({EntryType::LOG_ELEMENT, element}))
-        LOG_ERROR(log, "SystemLog queue is full");
+    if (queue.size() >= DBMS_SYSTEM_LOG_QUEUE_SIZE / 2)
+    {
+        // The queue more than half full, time to flush.
+        const uint64_t queue_end = queue_front_index + queue.size();
+        if (requested_flush_before < queue_end)
+        {
+            requested_flush_before = queue_end;
+        }
+        flush_event.notify_all();
+    }
+
+    if (queue.size() >= DBMS_SYSTEM_LOG_QUEUE_SIZE)
+    {
+        // TextLog sets its logger level to 0, so this log is a noop and there
+        // is no recursive logging.
+        LOG_ERROR(log, "Queue is full for system log '" + demangle(typeid(*this).name()) + "'.");
+        return;
+    }
+
+    queue.push_back(element);
 }
 
 
 template <typename LogElement>
 void SystemLog<LogElement>::flush()
 {
+    std::unique_lock lock(mutex);
+
     if (is_shutdown)
         return;
 
-    std::lock_guard flush_lock(flush_mutex);
-    force_flushing = true;
+    const uint64_t queue_end = queue_front_index + queue.size();
 
-    /// Tell thread to execute extra flush.
-    queue.push({EntryType::FORCE_FLUSH, {}});
+    if (requested_flush_before < queue_end)
+    {
+        requested_flush_before = queue_end;
+    }
+    flush_event.notify_all();
 
-    /// Wait for flush being finished.
-    std::unique_lock lock(condvar_mutex);
-    while (force_flushing)
-        flush_condvar.wait(lock);
+    const int timeout_seconds = 60;
```

Comment:
This is not a meaningful parameter, just some arbitrary timeout to avoid endless waits. I hope that someday we switch to our own mutex to eliminate all endless waits from our code, and its timeout cap will be configurable.